### PR TITLE
Lint all head changes in one process

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

On pre-push, committed changes to files are linted concurrently by piping them them into `eslint`. The benefit of this approach is that it allows us to ignore uncommitted "dirty" changes, as these changes are not being pushed. 

The trouble with this approach is that a different process is generated for every file. If a lot of files have been changed, especially after a large rebase, it caused too many concurrent processes to spawn, and node to run out of memory.

This new approach is to lint all files in one process, avoiding the memory issue. However it does mean that uncommitted changes to _files that were previously changed and committed on the branch_ are linted. This introduces an edge case whereby changes that would not be pushed but fail linting prevent all code being pushed.  

I have tried several different approaches, this one having the least friction. I'm open to any ideas!

## What is the value of this and can you measure success?

Validation does not blow up and prevent user from pushing a branch

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
Yeah! I think you can have the best of both worlds through batching, something like this:
```
const batchSize = 10;
...
batch(jsFiles, batchSize).reduce(lint, Promise.resolve());
function batch(arr, size) {
  function batch_fold(xss, x) {
    if( !xss.length ) return [[x]];
    if( xss[0].length < batchSize ) return [xss[0].concat(x), ...xss.slice(1)];
    return [[x], ...xss];
  }
  return arr.reduce(batch_fold, []);
}
function lint(proc, batch) {
  return proc.then(Promise.all(...));
}
```

File: tools/__tasks__/validate-head/javascript.js [hunk 1]
```diff
@@ -9,23 +9,48 @@ module.exports = {
...
-                                file === 'git-hooks/pre-push'
-                        )
-                        .join(' ');
+                    const jsFiles = files.filter(
+                        file =>
+                            file.endsWith('.js') ||
+                            file === 'git-hooks/pre-push'
+                    );
+                    const
+                        lint = proc =>
```
Comment (regiskuckaertz) [inline tools/__tasks__/validate-head/javascript.js]: Second arg missing here: `jsFiles`. Will be passed by `reduce`.

**Comment by regiskuckaertz** [comment]
Aside from missing arg in reducer, lgtm 👍  AWESOME!

**Comment by gustavpursche** [comment]
Nice work! 🍻

**Comment by SiAdcock** [comment]
Thanks so much @regiskuckaertz, I'm much happier with this approach! 

I've tweaked this a little bit so that the reducer doesn't fail entirely if linting fails for a particular file. Instead we accumulate errors, reduce the `stdout`, concatenate with a nice helpful error message and finally reject.

File: tools/__tasks__/validate-head/javascript.js [hunk 2]
```diff
@@ -1,7 +1,10 @@
 const execa = require('execa');
 const chalk = require('chalk');
+const os = require('os');
 const getChangedFiles = require('../lib/get-changed-files');
 
+const getCpuCount = () => os.cpus().length;
```
Comment (regiskuckaertz) [inline tools/__tasks__/validate-head/javascript.js]: 👏

Comment (SiAdcock) [reply tools/__tasks__/validate-head/javascript.js]: A @gustavpursche innovation 🥇

**Review by gu-stav — APPROVED** [review]
🚅

**Pull request merged**